PIPS/pip-2.md

```diff
@@ -0,0 +1,35 @@
+---
+pip: 2
+title: Relayer Fees
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-30
+---
+
+## Simple Summary
+The current LockProxy implementation does not have an in-built method for relayers to charge fees for their work.
+
+This PIP proposes a simple method to charge fees.
+
+## Motivation
+Network fees can very expensive. For example, the total Ethereum network fees for some popular DApps can cost up to 500k USD over 30 days (source: https://ethgasstation.info/).
+
+## Specification
+The `lock` and `unlock` function of the LockProxy can be modified to include the following parameters:
+- bytes memory feeReceiverAddr
+- uint256 feeAmount
+- uint64 feeChainId
+
+The `feeAmount` indicates how much of the locked `amount` should be deducted and given to the `feeReceiverAddr`. The `feeChainId` allows control over which chain the fee deduction should be performed at.
+
+Example fee deduction implementation:
+```
+uint256 remainingAmount = amount;
+
+if (feeAmount > 0 && self.chainId == feeChainId) {
+  remainingAmount -= feeAmount;
```

Comment:
It comes to me first that the feeAmount should be separated from amount. Suppose I want to transfer some Ether from account Eth_A wants to Ontology account Ont_A with fee of 0.1 Ether, I would designate the amount to be transferred is, say, 10 Ether, then intuitively, I expect my Ont_A account will receive 10 Ether, rather than 9.9 Ether. 
So I'm thinking the fee amount should be the additional value rather than deducted directly from the amount.

Reply:
Hmm usually for us, we prefer to deduct the fee amount from the amount, because we find it easier to think about.
For example, if the user only has 10 Ether and wants to send everything to Neo, he makes a request to the relayer to ask how much he should pay for the fee, and the relayer returns an amount of 0.0173 Ether.

If the fee amount is separate from the amount, then the user has to calculate and change his amount to 9.9827 before sending.

Reply:
Hm, anyway, it doesn't matter much, I think.
Here is another doubt. Will feeChainId always be toChainId or targetChainId, or can also be fromChainId?
If feeChainId is set as fromChainId, then there is no way to (currently) prevent relayer receiving the fee from not relaying from PolyChain to toChain, even not relaying tx from fromChain to PolyChain.

Reply:
I think we could allow the user / relayer to have the flexibility to set feeChainId as toChainId or fromChainId.
For example, a DApp can have relayers with some reputation and the user could trust the relayer.

The advantage of having it flexible is that it would help the relayer to not end up with dust amounts on two different chains.

Reply:
I think the fee amounts should be quite small, so it is not worth cheating a few users and then losing their own reputation

Reply:
Then it comes to another question. How to calculate and record the reputation?  And then there should be a mechanism to trace the users' trusted relayer account address which relays the cross chain tx and monitor the relayers' behavior. 
However, if it comes to the case where users trust the relayer, we don't need to worry at all. Maybe this point should be included in proposal.
In addition, how do say if we add another parameter named feeAssetHash? Currently the fee is type of asset being crossed by default.

Reply:
I think recording the reputation can be left to the DApp.
Makes sense, I will add that point to the proposal.

I think feeAssetHash is a good idea, will add it in.